money-java-servlet/src/main/scala/com/comcast/money/java/servlet/TraceFilter.scala

```diff
@@ -16,52 +16,94 @@
  */
 class TraceFilter extends Filter {
 
   private val MoneyTraceHeader = "X-MoneyTrace"
+  private val B3TraceId = "X-B3-TraceId"
+  private val B3SpanId = "X-B3-SpanId"
+  private val B3ParentSpanId = "X-B3-ParentSpanId"
   private val logger = LoggerFactory.getLogger(classOf[TraceFilter])
   private val factory = Money.Environment.factory
 
   override def init(filterConfig: FilterConfig): Unit = {}
 
   override def destroy(): Unit = {}
 
+  private val spanName = "servlet"
+
   override def doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = {
 
     SpanLocal.clear()
     val httpRequest = new HttpServletRequestWrapper(request.asInstanceOf[HttpServletRequest])
-    val incomingTraceId = Option(httpRequest.getHeader(MoneyTraceHeader)) map { incTrcaceId =>
-      // attempt to parse the incoming trace id (its a Try)
-      fromHttpHeader(incTrcaceId) match {
-        case Success(spanId) => SpanLocal.push(factory.newSpan(spanId, "servlet"))
-        case Failure(ex) => logger.warn("Unable to parse money trace for request header '{}'", incTrcaceId)
+
+    val spanId: Option[SpanId] = parseMoneyTraceHeader(httpRequest).orElse(parseB3TraceHeader(httpRequest))
+    spanId.foreach(s => SpanLocal.push(factory.newSpan(s, spanName)))
+
+    val httpResponse = response.asInstanceOf[HttpServletResponse]
+    setResponseHeader(MoneyTraceHeader)
+    setResponseHeader(B3TraceId)
+    setResponseHeader(B3ParentSpanId)
+    setResponseHeader(B3SpanId)
+
+    def setResponseHeader(headerName: String) = {
+      Option(httpRequest.getHeader(headerName)).foreach(v => httpResponse.addHeader(headerName, v))
+    }
+    chain.doFilter(request, response)
+  }
+
+  def parseMoneyTraceHeader(httpRequest: HttpServletRequest): Option[SpanId] = {
+
+    def parseHeader(headerValue: String): Option[SpanId] = {
+      fromHttpHeader(headerValue) match {
+        case Success(spanId) => Some(spanId)
+        case Failure(ex) => {
+          logger.warn("Unable to parse money trace for request header '{}'", headerValue)
+          None
+        }
       }
-      incTrcaceId
     }
 
-    incomingTraceId.foreach { traceId =>
-      response match {
-        case http: HttpServletResponse =>
-          http.addHeader(MoneyTraceHeader, traceId)
-        case _ =>
-          logger.warn("Unable to set money trace header on response, response type is not an HttpServletResponse ")
+    Option(httpRequest.getHeader(MoneyTraceHeader)) match {
+      case Some(x) => parseHeader(x)
+      case _ => None
+    }
+  }
+
+  def parseB3TraceHeader(httpRequest: HttpServletRequest): Option[SpanId] = {
+
+    def parseHeaders(traceIdVal: String): Option[SpanId] = {
+      val maybeParentSpanId = Option(httpRequest.getHeader(B3ParentSpanId))
+      val maybeSpanId = Option(httpRequest.getHeader(B3SpanId))
+      fromB3HttpHeaders(traceIdVal, maybeParentSpanId, maybeSpanId) match {
```

Comment:
fromB3HttpHeaders is in core, which currently has no dependency on javaxServlet. javaxServlet has HttpServletRequest defined, so core would need a dependency on javaxServlet...which does not seem right.

fromB3HttpHeaders is in core because money-http-client also has a dependency on it

Reply:
More correctly money-http-client has a dependency on  Formatters, the containing object.

Reply:
Haha yeah. I see that now. Good call.